src/common/common.ts

```diff
@@ -109,6 +119,43 @@ export function isVerifiableCredential(
   return dataIsVc;
 }
 
+function isUrl(url: string): boolean {
+  try {
+    // If url is not URL-shaped, this will throw.
+    // eslint-disable-next-line no-new
+    new URL(url);
+    return true;
+  } catch (_e) {
+    return false;
+  }
+}
+
+export function isVerifiablePresentation(
+  vp: unknown | VerifiablePresentation
+): vp is VerifiablePresentation {
+  let inputIsVp = true;
+  inputIsVp = inputIsVp && Array.isArray((vp as VerifiablePresentation).type);
+  if ((vp as VerifiablePresentation).verifiableCredential !== undefined) {
+    inputIsVp =
+      inputIsVp &&
+      Array.isArray((vp as VerifiablePresentation).verifiableCredential);
+    inputIsVp =
+      inputIsVp &&
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      (vp as VerifiablePresentation).verifiableCredential!.every(
+        isVerifiableCredential
+      );
+  }
+  if ((vp as VerifiablePresentation).holder !== undefined) {
+    inputIsVp =
+      inputIsVp && typeof (vp as VerifiablePresentation).holder === "string";
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    inputIsVp = inputIsVp && isUrl((vp as VerifiablePresentation).holder!);
+  }
+  // TODO: No type checking is currently implemented for the proof.
+  return inputIsVp;
+}
```

Comment:
```suggestion
function isUnknownObject(x: unknown): x is {
  [key in PropertyKey]: unknown;
} {
  return x !== null && typeof x === "object";
}

export function isVerifiablePresentation(
  x: unknown | VerifiablePresentation
): x is VerifiablePresentation {
  // TODO: No type checking is currently implemented for the proof.
  return (
    isUnknownObject(x) &&
    Array.isArray(x.type) &&
    Array.isArray(x.verifiableCredential) &&
    x.verifiableCredential.every(isVerifiableCredential) &&
    typeof x.holder === "string" &&
    isUrl(x.holder)
  );
}
```

Reply:
I would suggest we start adopting this style of type assertions because it is much easier to read ^^,

Reply:
I'd prefer to keep it consistent within the codebase for now. Also, I realized that the reason why I have casts everywhere is because of the multi-stage nature of the approach, and not a single chained expression. It's easier to debug, since you can put breakpoints on each step of the type guard, so I'm really hesitant to switch away from this.

Reply:
<img width="948" alt="Screenshot 2021-10-01 at 10 33 26" src="https://user-images.githubusercontent.com/613619/135598911-24d073a9-ca35-46e8-ba50-7f5cd60c2613.png">

You can do breakpoint on any line with that style! ;)

Reply:
Ah my bad, I assumed it could only break on a single expression. I don't mind the refactoring then, as long as the codebase remains consistent :)

Reply:
But wait, the semantics of the expression is changed in your proposal: for instance, `holder` is either undefined, either an IRI-shaped string. Wouldn't that make the expression more cumbersome ?

Reply:
Haha, yes, I noticed ^^,

You could do something like that:
```typescript
export function isVerifiablePresentation(
  x: unknown | VerifiablePresentation
): x is VerifiablePresentation {
  // TODO: No type checking is currently implemented for the proof.
  return (
    isUnknownObject(x) &&
    Array.isArray(x.type) &&
    ("verifiableCredential" in x
      ? Array.isArray(x.verifiableCredential) &&
        x.verifiableCredential.every(isVerifiableCredential)
      : true) &&
    ("holder" in x ? typeof x.holder === "string" && isUrl(x.holder) : true)
  );
}
```

Or possibly create more specific or utility functions to check for things that can be undefined.

Agreed there is a bit more work to do on type guards to make them more readable even in this form.